### PR TITLE
Improve performance of visual mode with many Ace instances

### PR DESCRIPTION
**Pull request by jmcphers**
This change significantly increases the performance of the visual editor when it contains many code chunks, both when initially opening the editor and when switching tabs to an existing editor.

The primary technique is to replace the per-editor `FontMetrics` object with a single `FontMetrics` objects shared among all editors in a document (see here for some rationale: https://github.com/ajaxorg/ace/issues/2153). 

However, doing this has a bunch of downstream side effects that are complicated to mitigate. The worst of these is that it turns out that the `FontMetrics` object is responsible for triggering the editor to redraw itself when the size of its container changes. Hoisting `FontMetrics` out of the editor requires us to reimplement this trigger. 

It also just takes more time to render many dozens of Ace instances than we'd like; even when we are only redrawing them when necessary, the redraw of > 100 editors can make the UI hang for several seconds. So we also need to defer redraws so that the UI is responsive while we the work happens. Luckily, most of these redraws are offscreen, so we use a sorted priority queue to ensure that the Ace instances that are onscreen get rendered first.

### CODE REVIEW

**Comment by jmcphers** [comment]
> Is this the destruction of the entire visual editor or the destruction of the ace editor? If the latter, how does the `fontMeasurer_` get re-created if new chunks are added? (forgive me if this is obvious based on better understanding of the code, just looking at a diff here)

I think that may have been true in the WIP commit but should be clean now (LMK if it still looks wrong!)

**Comment by jjallaire** [comment]
Wow, quite the odyssey to figure all of that out! This LGTM as far as my knowledge of Ace and our embedding of it goes. Will go ahead and merge.

File: src/gwt/acesupport/loader.js [hunk 1]
```diff
@@ -253,33 +254,192 @@ oop.inherits(RStudioUndoManager, UndoManager);
...
+
+      // typically the font metrics change is what triggers a redraw when the
+      // size becomes nonzero; this is needed so that e.g. an editor that was
+      // initially rendered in an unselected tab can re-render itself with the
+      // correct metrics when it becomes visible.
+      //
+      // without font the font metrics system, we need another way to trigger
+      // that redraw, so we implement a direct resize observer here.
+      self.$cachedHeight = 0;
+      self.$resizeObserver = new window.ResizeObserver(function(e) {
```
Comment (kevinushey) [inline src/gwt/acesupport/loader.js:349]: Are these APIs available on IE11? (Do we need to worry about IE11 for visual mode, or will we drop support for that in time for the 1.4 release?)

**Comment by jjallaire** [comment]
Doing some early testing on this and I'm wondering whether the render queue is something we absolutely need. A couple reasons for this:

1) I've noticed that this re-introduces a behavior that CodeMirror had where the height of the document is fluid when it's first opened (as the various editors measure themselves). This makes navigation to targets within the file non-deterministic when the file must be opened as part of the navigation (the main way navigation will happen is though cross-reference links, which are provided in an inline popup as well as now in omni-search). I realize that without the render-queue the MANUAL.md file is painful to open, but perhaps we're willing to trade some very, very long documents being slow to load for more predictable/consistent navigation behavior?

2) Of more concern, I did a test drive of a project I use for more realistic loading scenarios (https://github.com/hadley/mastering-shiny). For the test I load all of the chapters in the book and switch them to visual mode, then reload the project. I think this is realistic b/c the low "cost" of opening a document tab encourages users to have a large working set of tabs. Anyway, in my first attempt at this many of the Ace instances within the chapter tabs never actually got a non-zero height:

Screen Shot 2020-07-22 at 5 40 42 AM

Nothing in the JS console to indicate errors. 

To me just having the risk of non-rendering be in play makes me quite hesitant. This did repro for me and perhaps it will repro for you and we can find a fix, but still worth a realtime discussion on the underlying cause and whether we want to keep on with the render queue.

**Comment by shaan6143** [comment]
unsubscribe

**Comment by jjallaire** [comment]
Another more serious problem introduced is that after you insert a new code chunk it appears drawn at ~ zero height.

**Comment by jjallaire** [comment]
Reverted the changes introduced by this PR here: https://github.com/rstudio/rstudio/commit/4f95b34a1d2d3d1c970c395c4a467b0275815405

**Pull request merged**